### PR TITLE
Task stream

### PR DESCRIPTION
**Pull request by ryanpbrewster**
I don't even really understand HOW the changes I made salvaged the memoization properties, but hallelujah

@leogrim

### CODE REVIEW

**Comment by leogrim** [comment]
Sweet, what did I do again? ;)

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  private case class Node[T](task: Task[(Option[T], SyncList[T])]) extends SyncList[T] {
+    override def foldLeftUntil[A](x0: A, p: Promise[A])(op: (A, T) => (A, Boolean))(implicit exc: ScalaExecutionContext): Task[A] = new Task(() => {
+      task.run.onComplete {
+        case Failure(fail) => p.failure(fail)
+        case Success((None, next)) =>
+          next.foldLeftUntil(x0, p)(op).run
+        case Success((Some(v), next)) =>
+          val (x1, done) = op(x0, v)
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala:73]: You may want a `Try` in here

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala [hunk 2]
```diff
@@ -0,0 +1,105 @@
+package com.keepit.common.concurrent
+
+import scala.concurrent.{ ExecutionContext => ScalaExecutionContext, Promise, Future }
+import com.keepit.common.core.futureExtensionOps
+
+import scala.util.{ Failure, Success }
+
+class Task[T](f: () => Future[T]) {
+  lazy val run: Future[T] = f()
```
Comment (leogrim) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala:9]: And you probably want a `Try` in here as well

File: kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala (hunk 1 above)
Comment (ryanpbrewster) [inline kifi-backend/modules/common/app/com/keepit/common/concurrent/SyncList.scala:73]: I think if I put the `Try` in the definition of `Task` that should take care of this as well, right?

**Pull request closed without merging**